share/wake/lib/system/remote_cache_api.wake

```diff
@@ -51,5 +171,368 @@ export def makeRemoteCacheApi (config: String): Result RemoteCacheApi Error =
+
+    require Pass response =
+        makeRoute api "blob"
+        | buildHttpRequest
+        | setMethod HttpMethodPost
+        | addAuthorizationHeader auth
+        | setBlob
+        | makeRequest
+
+    require (Some 200) = response.getHttpResponseStatusCode
+    else
+        failWithError
+        "rsc: failed to upload string as blob with code {response.getHttpResponseStatusCode | format}"
+
+    require Pass json = response.getHttpResponseBody.parseJSONBody
+
+    require Pass (JString "Ok") = jField json "type"
+    else failWithError "rsc: failed to upload blob"
+
+    require Pass (JArray blobs) = jField json "blobs"
+    else failWithError "rsc: JSON response has incorrect schema for blob upload. (Missing blobs key)"
+
+    def idJVals =
+        blobs
+        | map (jField _ "id")
+
+    require Pass ids = idJVals.findFail
+    else failWithError "rsc: JSON response has incorrect schema for blob upload. (Missing blob ids)"
+
+    require Some (JString id) = ids.head
+    else
+        failWithError
+        "rsc: JSON response has incorrect schema for blob upload. (Expected only 1 id saw {ids.len.str})"
+
+    Pass id
+
+# Converts a CacheSearchRequest to JSON
+def getCacheSearchRequestJson (req: CacheSearchRequest): JValue =
+    def (CacheSearchRequest cmd cwd env hiddenInfo isAtty stdin visibleFiles) = req
+
+    JObject (
+        "cmd" :-> JArray cmd.asBytesDelimedByNull.toJIntegerList,
+        "cwd" :-> JString cwd,
+        "env" :-> JArray env.asBytesDelimedByNull.toJIntegerList,
+        "hidden_info" :-> JString hiddenInfo,
+        "is_atty" :-> JBoolean isAtty,
+        "stdin" :-> JString stdin,
+        "visible_files" :-> JArray (visibleFiles | map getPathAsJson),
+    )
+
+# Converts a CachePostJobRequest to JSON
+def getCachePostRequestJson (req: CachePostRequest): JValue =
+    def (
+        CachePostRequest
+        cmd
+        cwd
+        env
+        hiddenInfo
+        isAtty
+        stdin
+        visibleFiles
+        # TODO: these two are just being ignored right now
+        # That is probably not right
+        _outputDirs
+        _outputSymlinks
+        outputFiles
+        stdoutBlobId
+        stderrBlobId
+        status
+        runtime
+        cputime
+        memory
+        ibytes
+        obytes
+    ) = req
+
+    def mkOutputFileJson ((CachePostRequestOutputFile path mode blobId): CachePostRequestOutputFile): JValue =
+        JObject (
+            "path" :-> JString path,
+            "mode" :-> JInteger mode,
+            "blob_id" :-> JString blobId,
+        )
+
+    JObject (
+        "cmd" :-> JArray cmd.asBytesDelimedByNull.toJIntegerList,
+        "cwd" :-> JString cwd,
+        "env" :-> JArray env.asBytesDelimedByNull.toJIntegerList,
+        "hidden_info" :-> JString hiddenInfo,
+        "is_atty" :-> JBoolean isAtty,
+        "stdin" :-> JString stdin,
+        "visible_files" :-> JArray (visibleFiles | map getPathAsJson),
```

Comment:
Generally I'm pro DRYing things out, but when it comes to JSON interfaces to external systems I rather keep serialization/deserialization independent since they really are separate types to the server and they can arbitrarily change in the future. If we merge them here and something changes then we have to uncouple everything again.

(aside this is why I want the json generator because all of this boilerplate would be generated/hidden)